book/src/asm/instructions.md

```diff
@@ -20,10 +20,54 @@ Instructions feature:
 
 # External instructions
 
-An external instruction delegates calls to a function inside a submachine of this machine. When it is called, a call is made to the submachine function. An example of an external instruction is the following:
+An external instruction delegates its implementation to a function/operation from a submachine.
+When called, the inputs and outputs of the declared instruction are mapped into a call to the submachine function.
 
+Assume we have a submachine with a single operation `add`:
 ```
-instr assert_zero X = my_submachine.assert_zero // where `assert_zero` is a function defined in `my_submachine`
+machine SubMachine(latch, operation_id) {
+    operation add<0> a,b -> c;
+    ...
+}
+```
+
+An external instruction calling into this operation can be declared as follows:
+```
+machine Main {
+    ...
+    reg X[<=];
+    reg Y[<=];
+    reg Z[<=];
+
+    instr add X, Y -> Z = submachine.add; // will call submachine.add(X, Y) and return through Z
+    // this is equivalent to:
+    // instr add X, Y -> Z = submachine.add X, Y -> Z;
+    ...
+}
+```
+The left-hand side of the definition declares the instruction parameters, that is, how the instruction is called in the local machine.
+The right-hand side of the definition declares the external operation being called, and how the instruction parameters are used in that call.
```

Comment:
```suggestion
The right-hand side of the definition specifies  the external operation being called, and how the instruction parameters are used in that call.
```

Reply:
Hm, I fear this still does not really catch the essence...
But I also don't have a better suggestion right now.

Reply:
Maybe something along the lines of "left hand side defines the name of the local instruction and and its parameters, who all need to be assignment registers. The right hand side specifies the name of the sub-machine instruction and how to call it. Here, in addition to the assignment registers used on the left hand side, non-assignment registers can be used. Those can be seen as implicit parameters."

Reply:
reworked the text and expanded the example